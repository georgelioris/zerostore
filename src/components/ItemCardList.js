import React from "react";
import { Image } from "./Image";

const ItemCard = ({ item, onItemClick }) => (
  <div className="card hoverable" key={item.id}>
    <div className="card-image">
      <Image url={item.img} width={"300px"} height={"200px"} />
      <span className="card-title">{item.title}</span>
      <span
        className={`btn-floating halfway-fab waves-effect waves-light red sidenav-trigger
          ${item.available === false ? " unavailable disabled" : ""}`}
        data-target="slide-out"
        onClick={onItemClick}
      >
        <span>
          <i className="material-icons">add</i>
        </span>
      </span>
    </div>

    <div className="card-content">
      <p>{item.desc}</p>
      <p>
        <b>Price: {item.price}$</b>
      </p>
    </div>
  </div>
);

const ItemCardList = ({ visibleItems, handleItemClick, cartItems }) => (
  <div>
    <div className="item-card-list">
      {visibleItems.map(item => (
        <ItemCard
          key={item.id}
          item={item}
          onItemClick={() => handleItemClick(item, cartItems)}
        />
      ))}
    </div>
  </div>
);

export default ItemCardList;
