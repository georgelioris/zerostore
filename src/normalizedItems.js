import ItemImg from "./images/swshoer.png";

export const defaultItems = {
  1: {
    id: 1,
    title: "winter body",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 110,
    img: ItemImg,
    category: "winter",
    available: true
  },
  2: {
    id: 2,
    title: "Adidas",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 80,
    img: ItemImg,
    category: "new",
    available: true
  },
  3: {
    id: 3,
    title: "Vans",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 120,
    img: ItemImg,
    category: "on sale",
    available: true
  },
  4: {
    id: 4,
    title: "White",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 260,
    img: ItemImg,
    category: "on sale",
    available: true
  },
  5: {
    id: 5,
    title: "Cropped-shoe",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 160,
    img: ItemImg,
    category: "summer",
    available: true
  },
  6: {
    id: 6,
    title: "Blues",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 90,
    img: ItemImg,
    category: "summer",
    available: true
  },
  7: {
    id: 7,
    title: "Black",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 390,
    img: ItemImg,
    category: "winter",
    available: true
  },
  8: {
    id: 8,
    title: "Another One",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 40,
    img: ItemImg,
    category: "new",
    available: true
  },
  9: {
    id: 9,
    title: "Orange",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 190,
    img: ItemImg,
    category: "new",
    available: true
  },
  10: {
    id: 10,
    title: "Model One",
    desc:
      "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.",
    price: 55,
    img: ItemImg,
    category: "new",
    available: true
  }
};
