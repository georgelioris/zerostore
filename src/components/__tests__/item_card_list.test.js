import React from "react";
import ItemCardList from "../ItemCardList";
import { unmountComponentAtNode } from "react-dom";
import { act, create } from "react-test-renderer";
import { defaultItems as items } from "../../normalizedItems";
import { BrowserRouter as Router } from "react-router-dom";

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

// Mock state and function
const mock = jest.fn();
const props = {
  visibleItems: Object.values(items).slice(0, 4),
  cartObject: { 1: { id: 1, quantity: 1 }, 4: { id: 4, quanity: 2 } },
  handleItemClick: mock
};

// Tests
describe("ItemCardList component", () => {
  const container = create(
    <Router>
      <ItemCardList {...props} />
    </Router>
  );
  it("Should match the snapshot", () => {
    expect(container.toJSON()).toMatchSnapshot();
  });
  it("Should dispatch handleItemClick on button click", () => {
    // Click all hadndleItemClick buttons
    act(() => {
      container.root
        .findAllByProps({
          "data-target": "slide-out"
        })
        .map(i => i.props.onClick());
    });
    expect(props.handleItemClick).toHaveBeenCalledTimes(4);
    expect(props.handleItemClick).toHaveBeenNthCalledWith(
      1,
      items[1],
      props.cartObject
    );
    expect(props.handleItemClick).toHaveBeenNthCalledWith(
      2,
      items[2],
      props.cartObject
    );
    expect(props.handleItemClick).toHaveBeenNthCalledWith(
      3,
      items[3],
      props.cartObject
    );
    expect(props.handleItemClick).toHaveBeenNthCalledWith(
      4,
      items[4],
      props.cartObject
    );
  });
  it("Should include Link to /shop/items/ID", () => {
    // Get all Link.to locations
    const links = container.root
      .findAllByProps({ className: "red-text" })
      .reduce((acc, i) => (i.props.to ? [...acc, i.props.to] : acc), []);
    expect(links).toStrictEqual([
      "/shop/items/1",
      "/shop/items/2",
      "/shop/items/3",
      "/shop/items/4"
    ]);
  });
});
