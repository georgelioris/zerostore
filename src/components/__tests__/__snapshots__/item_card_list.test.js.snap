// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCardList component Should match the snapshot 1`] = `
<div>
  <div
    className="item-card-list"
  >
    <div
      className="row"
    >
      <div
        className="col s12 m6"
      >
        <div
          className="card"
        >
          <div
            className="card-image"
          >
            <div
              className="image-component"
              style={
                Object {
                  "background": "url(swshoer.png)",
                  "height": "200px",
                  "width": "300px",
                }
              }
            />
            <div
              className="btn-floating halfway-fab waves-effect waves-light red accent-2  sidenav-trigger"
              data-target="slide-out"
              onClick={[Function]}
            >
              <span>
                <i
                  className="material-icons"
                >
                  add
                </i>
              </span>
            </div>
          </div>
          <div
            className="card-content"
          >
            <span
              className="card-title"
            >
              Winter body
            </span>
            <p
              className="truncate"
            >
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.
            </p>
            <p>
              <b>
                Price: 
                110
                $
              </b>
            </p>
          </div>
          <div
            className="card-action grey lighten-4"
          >
            <button
              className="waves-effect red accent-1 white-text waves-light  btn-flat"
              onClick={[Function]}
            >
              winter
            </button>
            <a
              className="red-text"
              href="/shop/items/1"
              onClick={[Function]}
            >
              More
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col s12 m6"
      >
        <div
          className="card"
        >
          <div
            className="card-image"
          >
            <div
              className="image-component"
              style={
                Object {
                  "background": "url(swshoer.png)",
                  "height": "200px",
                  "width": "300px",
                }
              }
            />
            <div
              className="btn-floating halfway-fab waves-effect waves-light red accent-2 unavailable disabled sidenav-trigger"
              data-target="slide-out"
              onClick={[Function]}
            >
              <span>
                <i
                  className="material-icons"
                >
                  add
                </i>
              </span>
            </div>
          </div>
          <div
            className="card-content"
          >
            <span
              className="card-title"
            >
              Purple Runner
            </span>
            <p
              className="truncate"
            >
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.
            </p>
            <p>
              <b>
                Price: 
                80
                $
              </b>
            </p>
          </div>
          <div
            className="card-action grey lighten-4"
          >
            <button
              className="waves-effect red accent-1 white-text waves-light  btn-flat"
              onClick={[Function]}
            >
              new
            </button>
            <a
              className="red-text"
              href="/shop/items/2"
              onClick={[Function]}
            >
              More
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col s12 m6"
      >
        <div
          className="card"
        >
          <div
            className="card-image"
          >
            <div
              className="image-component"
              style={
                Object {
                  "background": "url(swshoer.png)",
                  "height": "200px",
                  "width": "300px",
                }
              }
            />
            <div
              className="btn-floating halfway-fab waves-effect waves-light red accent-2  sidenav-trigger"
              data-target="slide-out"
              onClick={[Function]}
            >
              <span>
                <i
                  className="material-icons"
                >
                  add
                </i>
              </span>
            </div>
          </div>
          <div
            className="card-content"
          >
            <span
              className="card-title"
            >
              Red Lows
            </span>
            <p
              className="truncate"
            >
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.
            </p>
            <p>
              <b>
                Price: 
                120
                $
              </b>
            </p>
          </div>
          <div
            className="card-action grey lighten-4"
          >
            <button
              className="waves-effect red accent-1 white-text waves-light  btn-flat"
              onClick={[Function]}
            >
              on sale
            </button>
            <a
              className="red-text"
              href="/shop/items/3"
              onClick={[Function]}
            >
              More
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col s12 m6"
      >
        <div
          className="card"
        >
          <div
            className="card-image"
          >
            <div
              className="image-component"
              style={
                Object {
                  "background": "url(swshoer.png)",
                  "height": "200px",
                  "width": "300px",
                }
              }
            />
            <div
              className="btn-floating halfway-fab waves-effect waves-light red accent-2  sidenav-trigger"
              data-target="slide-out"
              onClick={[Function]}
            >
              <span>
                <i
                  className="material-icons"
                >
                  add
                </i>
              </span>
            </div>
          </div>
          <div
            className="card-content"
          >
            <span
              className="card-title"
            >
              White
            </span>
            <p
              className="truncate"
            >
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.
            </p>
            <p>
              <b>
                Price: 
                260
                $
              </b>
            </p>
          </div>
          <div
            className="card-action grey lighten-4"
          >
            <button
              className="waves-effect red accent-1 white-text waves-light  btn-flat"
              onClick={[Function]}
            >
              on sale
            </button>
            <a
              className="red-text"
              href="/shop/items/4"
              onClick={[Function]}
            >
              More
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
